aidial_adapter_vertexai/chat/bison/prompt.py

```diff
@@ -17,31 +18,75 @@ def __repr__(self) -> str:
         return f"{self.value!r}"
 
 
-class BisonPrompt(BaseModel):
-    context: Optional[str]
-    messages: List[ChatMessage]
+class BisonPrompt(BaseModel, Truncatable):
+    context: Optional[str] = None
+    history: List[ChatMessage] = []
+    prompt: str
 
     @classmethod
     def parse(cls, messages: List[Message]) -> "BisonPrompt":
-        context, messages = _validate_messages_and_split(messages)
+        context, messages = _validate_and_split_messages(messages)
+        bison_messages = list(map(_parse_message, messages))
+
         return cls(
             context=context,
-            messages=list(map(_parse_message, messages)),
+            history=bison_messages[:-1],
+            prompt=bison_messages[-1].content,
         )
 
     @property
-    def user_prompt(self) -> str:
-        return self.messages[-1].content or ""
+    def has_system_message(self) -> bool:
+        return self.context is not None
 
-    @property
-    def history(self) -> List[ChatMessage]:
-        return self.messages[:-1]
+    def keep(self, index: int) -> bool:
+        # Keep the system message...
+        if self.context is not None and index == 0:
+            return True
+        index -= self.has_system_message
```

Comment:
Can you elaborate this a bit more?

Why do we shift index, if in `_validate_and_split_messages` we remove System message:
```python
    context: Optional[str] = None
    if len(messages) > 0 and messages[0].role == Role.SYSTEM:
        context = collect_text_content(messages[0].content)
        context = context if context.strip() else None
        messages = messages[1:] <----
```

Reply:
Because the index `idx` points to a message in a list of **all** messages including the system one:

```
message: user bot user bot user
index:   0    1   2    3   4
```

```
message: system user bot user bot user
index:   0      1    2   3    4   5
```

So when the system message is present, we need to account for it by adjusting the index to match its start with the start of the `history` list.

Reply:
Fixed, I simplified the condition for the last message.